### ISSUE TITLE
Unable to understand output mapping of YOLOv5 with number of grids and detection

### ISSUE THREAD

**Issue by aravindchakravarti** [ultralytics/yolov5#12479]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I understood the architecture of YOLOv5. I understand that we are doing detection at 3 scales. And number of grids will be equal to 
```
= 80x80x3 + 40x40x3 + 20x20x3 
= 25200
```

For each grid, will have 5+nc values. Assuming 80 classes, 
```
= 25200*85 
= 2,142,000 values
```
So, Yolo v5 should output ```2,142,000``` number of points. 

When I look at this line in detect.py (line 129)
```
  with dt[1]:
            visualize = increment_path(save_dir / Path(path).stem, mkdir=True) if visualize else False
            pred = model(im, augment=augment, visualize=visualize)
```

I checked type of 'pred'. It is an array of tensor.  
But, how do I know, in what way this 'pred' array packed 2,142,000 values?

I am not understanding the dimensionality of it.

 dim [0]     of  ```pred``` is ```torch.Size([1, 15120, 85])```
 dim [1] [0] of ```pred``` is ```torch.Size([1, 3, 48, 80, 85])```
 dim [1] [1] of ```pred``` is ```torch.Size([1, 3, 24, 40, 85])```
 dim [1] [2] of ```pred``` is ```torch.Size([1, 3, 12, 20, 85])```
 
I am not understanding how to link above tensor with 25200 grids

### Additional

Also, in ```non_max_suppression``` function I am seeing that, YOLO V5 is taking detection only from ```pred[0]``` which gives me a hint that, only we should worry only about ```pred[0]``` and not ```pred[1][:]```. Looks like pred[0] is ```inference_out``` and pred[1] is ```loss_out```. What are they actually?

```
    def non_max_suppression(
            prediction,
            conf_thres=0.25,
            iou_thres=0.45,
            classes=None,
            agnostic=False,
            multi_label=False,
            labels=(),
            max_det=300,
            nm=0,  # number of masks
    ):
    ...
    ...
    # Checks
    assert 0 <= conf_thres <= 1, f'Invalid Confidence threshold {conf_thres}, valid values are between 0.0 and 1.0'
    assert 0 <= iou_thres <= 1, f'Invalid IoU {iou_thres}, valid values are between 0.0 and 1.0'
    if isinstance(prediction, (list, tuple)):  # YOLOv5 model in validation model, output = (inference_out, loss_out)
        prediction = prediction[0]  # select only inference output
````

**Comment by glenn-jocher** [comment]
@aravindchakravarti the model output `pred` has dimensions `[1, 15120, 85]` which corresponds to the detection predictions for each grid cell across all scales. The 15120 value comes from 80x80x3 + 40x40x3 + 20x20x3. The `[1, 3, 48, 80, 85]`, `[1, 3, 24, 40, 85]`, and `[1, 3, 12, 20, 85]` are intermediate tensors representing predictions at each scale.

In `non_max_suppression` function, YOLOv5 selects only the inference output `pred[0]` for post-processing.

The `pred[1]` is the loss output representing the model training loss and isn't used for inference.

For further details, you can refer to the model's documentation at https://docs.ultralytics.com/yolov5/.

**Comment by aravindchakravarti** [comment]
@glenn-jocher Thanks Glenn, for the quick reply.
Understood that, `[1, 3, 48, 80, 85]`, `[1, 3, 24, 40, 85]`, and `[1, 3, 12, 20, 85]` are intermediate tensors representing loss.

One more small doubt, `80x80x3 + 40x40x3 + 20x20x3 = 25200` Right?
Why we have `15120` as the output from `pred[0]`. 

Sorry, I didn't understand the relationship between `25200` and `15120`

But if I do, `48*80*3 + 24*40*3 + 12*20*3 = 15120` !!!

So, are we creating grid of `48x80`, `24x40` and `12x20` instead of `80x80`, `40x40` and `20x20` ? If so why? Because of aspect ratio of image, correct?

**Comment by aravindchakravarti** [comment]
@glenn-jocher Looks like my above understanding is correct. If I feed square image (1:1 aspect ratio) i am getting `pred[0] = [1,25200,85]`
Also, `pred[1]` is `[1, 3, 80, 80, 85]`, `[1, 3, 40, 40, 85]`, and `[1, 3, 20, 20, 85]`

I understood now. 

If I have any more questions, I will get back to you.

Thank you once again!

**Comment by glenn-jocher** [comment]
@aravindchakravarti you got it right! The grid sizes in `pred[1]` reflect the aspect ratio of the input image, hence the values `[1, 3, 80, 80, 85]`, `[1, 3, 40, 40, 85]`, and `[1, 3, 20, 20, 85]` for different aspect ratios.

Feel free to reach out if you have any more questions. Happy to help!
